Support ”cherry-picked” requires

timriley opened the pull request:
This PR will make it so that any individual file can be required and work as expected (e.g. `require "dry/monads/result"`).

To avoid circular requires, I've started with @flash-gordon's suggestion to make it so that each monad class only provides "stub" implementations of their conversion methods that raise helpful errors, e.g.

```ruby
require "dry/monads/result"

Dry::Monads.Success("hello").to_maybe
# RuntimeError: Load Maybe first with require 'dry/monads/maybe'
```

And the real implementation of each conversion method is provided by the file of the class that is the _target_ of the conversion:

```ruby
require "dry/monads/result"
require "dry/monads/maybe"

Dry::Monads.Success("hello").to_maybe
# => Some("hello")
```

TODO:

- [x] Extract all conversions
- [x] Add tests for errors raised in stub implementations

timriley commented on lib/dry/monads/result.rb:
```diff
@@ -107,11 +106,9 @@ def to_s
-          Dry::Monads::Maybe(@value)
-        end
+          raise "Load Maybe first with require 'dry/monads/maybe'"
+        end unless method_defined?(:to_maybe)
```
@flash-gordon You happy with the way I'm conditionally defining these methods? If `"dry/monads/maybe"` is required first I don't want it's helpful implementation of `Result::Success#to_maybe` to be overridden by this one here.

timriley commented:
My last few commits here demonstrate the other side of this work: declaring the top-level Dry::Monads  constructors/constant aliases inside each individual file (which I've demonstrated with result.rb so far).

And I've also added a spec that does the usual $LOADED_FEATURES/remove_const tricks to be able to verify that the conversion method stubs raise errors when their target monad file isn't required yet.

I'll expand this work to cover all the files shortly.

This leaves one big thing: how do we make it so `include Dry::Monads` can still work, to include them all? i.e. this bit:

```ruby
    def self.included(base)
      super

      # TODO: Fix once CONSTRUCTORS is empty
      base.include(*CONSTRUCTORS)
    end
```

Could we make a special `dry/monads/all.rb` which adds this behaviour? Any other ideas, @flash-gordon?

timriley commented:
@flash-gordon You happy with the approach I started in https://github.com/dry-rb/dry-monads/pull/58/commits/4ee57cf6c1ec3214065db61dd0395bd67b1bce69?

flash-gordon commented:
@timriley this looks fine 👍 
I somehow missed the email about `all.rb`, sorry about that. I like the approach, let's have `dry/monads/all.rb` which loads all the monads and defines the `included` hook. We can also have a default `included` with a warning which says "you need to load 'all' first in order to import constructors", wdyt?

timriley commented:
@flash-gordon finally sorted some things out and I think this is ready for review now.

There are 3 things I wanted to point out from my work on this in the last couple of days:

1. In `conversion_stubs_spec.rb`, I tried various approaches of simulating the library not yet being fully loaded (e.g. un-defining all constants under `Dry::Monads`), but that ended up wreaking havoc on the rest of the test suite, causing errors everywhere. The approach I settled on at the end is a lot more "surgical", e.g. directly undermining the method we want to test the stub for, and then re-requiring the file to add the method back afterwards so its available for the rest of the suite. I feel like this is a fair compromise – how does it feel to you?
2. In `dry/monads.rb`, I adjusted `.included` to check for the `CONSTRUCTORS` constant provided by `dry/monads/all.rb` to determine whether to include the constructors or raise an error. I did try to work out some kind of inheritance-based approach for providing this behaviour, but nothing sat very well given that `dry/monads/all` doesn't actually provide its own module.
3. Like (1), in `all_monads_inclusion_spec.rb`, instead of fully unloading things, I've just removed the `CONSTRUCTORS` const temporarily to test the behaviour of raising the error when `dry/monads/all` hasn't been required.

Since it's been quite a circuitous path getting here, once you're happy with things, I squish all the commits here and we can get this merged :)

flash-gordon commented:
yeah, this looks great to me, thank you @timriley gonna test it and push beta2

timriley commented:
Oh great, you’re the best, thanks!

> On 20 May 2018, at 10:25 am, Nikita Shilnikov <notifications@github.com> wrote:
> 
> @timriley I squashed them with GH bffa1b6
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
